drivers/lte_link_control/lte_lc.c

```diff
@@ -63,107 +64,66 @@ static const char cgdcont[] = "AT+CGDCONT="CONFIG_LTE_PDP_CONTEXT;
-	if (len != size) {
-		LOG_ERR("send: failed");
-		return -EIO;
-	}
-
-	len = recv(fd, buffer, LC_MAX_READ_LENGTH, 0);
-	if ((len < AT_CMD_SIZE(success)) ||
-	    (memcmp(success, buffer, AT_CMD_SIZE(success)) != 0)) {
-		LOG_ERR("recv: %s", buffer);
-		return -EIO;
+	if (!memcmp(status1, response, AT_CMD_SIZE(status1)) ||
+	    !memcmp(status2, response, AT_CMD_SIZE(status2)) ||
+	    !memcmp(status3, response, AT_CMD_SIZE(status3)) ||
+	    !memcmp(status4, response, AT_CMD_SIZE(status4))) {
+		link = true;
 	}
-
-	return 0;
 }
 
 static int w_lte_lc_init_and_connect(struct device *unused)
 {
-	int err;
-	int at_socket_fd;
-
-	u8_t buffer[LC_MAX_READ_LENGTH];
-
-	at_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
-	if (at_socket_fd == -1) {
-		return -EFAULT;
-	}
-
 #if defined(CONFIG_LTE_EDRX_REQ)
 	/* Request configured eDRX settings to save power */
-	err = at_cmd(at_socket_fd, edrx_req, AT_CMD_SIZE(edrx_req));
-	if (err) {
-		close(at_socket_fd);
-		return err;
+	if (at_cmd_write(edrx_req, NULL) != AT_CMD_OK) {
+		return -EIO;
 	}
 #endif
-	err = at_cmd(at_socket_fd, subscribe, AT_CMD_SIZE(subscribe));
-	if (err) {
-		close(at_socket_fd);
-		return err;
+	if (at_cmd_write(subscribe, NULL) != AT_CMD_OK) {
+		return -EIO;
 	}
 
 #if defined(CONFIG_LTE_LOCK_BANDS)
 	/* Set LTE band lock (volatile setting).
 	 * Has to be done every time before activating the modem.
 	 */
-	err = at_cmd(at_socket_fd, lock_bands, AT_CMD_SIZE(lock_bands));
-	if (err) {
-		close(at_socket_fd);
-		return err;
+	if (at_cmd_write(lock_bands, NULL) != AT_CMD_OK) {
+		return -EIO;
 	}
 #endif
 #if defined(CONFIG_LTE_LEGACY_PCO_MODE)
-	err = at_cmd(at_socket_fd, legacy_pco, AT_CMD_SIZE(legacy_pco));
-	if (err) {
-		close(at_socket_fd);
-		return err;
+	if (at_cmd_write(legacy_pco, NULL) != AT_CMD_OK) {
+		return -EIO;
 	}
 	LOG_INF("Using legacy LTE PCO mode...");
 #endif
 #if defined(CONFIG_LTE_PDP_CMD)
-	err = at_cmd(at_socket_fd, cgdcont, AT_CMD_SIZE(cgdcont));
-	if (err) {
-		close(at_socket_fd);
-		return err;
+	if (at_cmd_write(cgdcont, NULL) != AT_CMD_OK) {
+		return -EIO;
 	}
 	LOG_INF("PDP Context: %s", cgdcont);
 #endif
-	err = at_cmd(at_socket_fd, network_mode, AT_CMD_SIZE(network_mode));
-	if (err) {
-		close(at_socket_fd);
-		return err;
-	}
-	err = at_cmd(at_socket_fd, normal, AT_CMD_SIZE(normal));
-	if (err) {
-		close(at_socket_fd);
-		return err;
+	if (at_cmd_write(network_mode, NULL) != AT_CMD_OK) {
+		return -EIO;
 	}
 
-	while (true) {
-		int bytes;
+	at_cmd_set_handler(at_handler);
+
+	if (at_cmd_write(normal, NULL) != AT_CMD_OK) {
+		return -EIO;
+	}
 
-		bytes = recv(at_socket_fd, buffer, LC_MAX_READ_LENGTH, 0);
-		if (bytes) {
-			LOG_DBG("recv: %s", buffer);
-			if (!memcmp(status1, buffer, AT_CMD_SIZE(status1)) ||
-			    !memcmp(status2, buffer, AT_CMD_SIZE(status2)) ||
-			    !memcmp(status3, buffer, AT_CMD_SIZE(status3)) ||
-			    !memcmp(status4, buffer, AT_CMD_SIZE(status4))) {
-				break;
-			}
-		}
+	while (!link) {
```

Comment:
We should get a notification from the Modem when the link is up.

Reply:
Yes, that's the notification parsed by the at_handler that raises the link flag.

Reply:
My point is: don't we have primitives in the OS to do better than a busy loop?
Network registration can take a few minutes (especially on NB-IoT).

Reply:
Hmm, I could use a semaphore and then wait for it to be given by the handler forever.

Reply:
> Hmm, I could use a semaphore and then wait for it to be given by the handler forever.

I think this would be more clean and optimized.

Reply:
Fixed, thank you.